client/executor/wasmtime/src/imports.rs

```diff
@@ -57,41 +44,44 @@ where
 			)))
 		}
 
-		if name == "memory" {
-			memory_import_index = Some(index);
-			externs.push((index, resolve_memory_import(store, &import_ty, heap_pages)?));
-			continue
-		}
-
 		match import_ty.ty() {
 			ExternType::Func(func_ty) => {
-				pending_func_imports.insert(name.to_owned(), (index, import_ty, func_ty));
+				pending_func_imports.insert(name.to_owned(), (import_ty, func_ty));
 			},
 			_ =>
 				return Err(WasmError::Other(format!(
-					"host doesn't provide any non function imports besides 'memory': {}:{}",
+					"host doesn't provide any non function imports: {}:{}",
 					import_ty.module(),
 					name,
 				))),
 		};
 	}
 
-	let mut registry = Registry { store, externs, pending_func_imports };
-
+	let mut registry = Registry { linker, pending_func_imports };
 	H::register_static(&mut registry)?;
-	let mut externs = registry.externs;
 
 	if !registry.pending_func_imports.is_empty() {
 		if allow_missing_func_imports {
-			for (_, (index, import_ty, func_ty)) in registry.pending_func_imports {
-				externs.push((
-					index,
-					MissingHostFuncHandler::new(&import_ty)?.into_extern(store, &func_ty),
-				));
+			for (name, (import_ty, func_ty)) in registry.pending_func_imports {
+				let error = format!("call to a missing function {}:{}", import_ty.module(), name);
+				if func_ty.params().len() == 0 && func_ty.results().len() == 0 {
+					linker
+						.func_wrap("env", &name, move || -> Result<(), Trap> {
+							Err(Trap::new(error.clone()))
+						})
+						.expect("adding a missing import stub doesn't fail");
+				} else {
+					log::warn!("Missing import: '{}' {:?}", name, func_ty);
+					linker
+						.func_new("env", &name, func_ty.clone(), move |_, _, _| {
```

Comment:
Not in this case; this callback is doing dynamic dispatch, so essentially the arguments are passed as an array. And since it's supposed to trap we just ignore them altogether.